targets/unity-v2/source/Shared/Public/Pipeline/OneDSEventPipeline.cs

```diff
@@ -0,0 +1,352 @@
+    /// such as writings events, batching and sending over network.
+    /// The public methods of this class are meant to be thread-safe and can be called from multiple threads.
+    /// </summary>
+    public class OneDSEventPipeline : IEventPipeline, IDisposable
+    {
+        // A flag that indicates whether the pipeline is currently active (started)
+        // and accepts incoming data (events).
+        private volatile bool isActive;
+        private object isActiveLock = new object();
+
+        private OneDSEventPipelineSettings settings;
+
+        // External cancellation token is used to cancel the pipeline from outside code
+        private CancellationToken externalCancellationToken;
+
+        // Pipeline creates and uses its own internal cancellation token source (based on the external token)
+        // to be able to communicate a cancellation between its stages when the cancellation needs to happen
+        // inside the pipeline itself (for example, when an exception occurs). Cancellation in this case 
+        // is needed to unblock all stage threads from waiting on data in the buffers.
+        //
+        // More about C# pipelines, their cancellation and error handling see here:
+        // https://msdn.microsoft.com/en-us/library/ff963548.aspx
+        private CancellationTokenSource pipelineCancellationTokenSource;
+
+        // Buffers used in the pipeline
+        private BlockingCollection<IPlayFabEmitEventRequest> eventBuffer;
+        private BlockingCollection<TitleEventBatch> batchBuffer;
+        private BlockingCollection<PlayFabResult<WriteEventsResponse>> sendResultBuffer;
+
+        // Stages used in the pipeline
+        private IPipelineStage<IPlayFabEmitEventRequest, TitleEventBatch> batchingStage;
+        private IPipelineStage<TitleEventBatch, PlayFabResult<WriteEventsResponse>> sendingStage;
+
+        // references to individual stages of the pipeline task
+        private Task batchingTask; 
+        private Task sendingTask;
+
+        // The composition of all pipeline tasks including stages
+        private Task pipelineTask;
+
+        private ILogger logger;
+
+        public OneDSEventPipelineSettings Settings
+        {
+            get
+            {
+                return this.settings;
+            }
+        }
+
+        public Task PipelineTask
+        {
+            get
+            {
+                return this.pipelineTask;
+            }
+        }
+
+        /// <summary>
+        /// Creates an instance of EventPipeline.
+        /// </summary>
+        /// <param name="settings">The configuration settings for event pipeline.</param>
+        public OneDSEventPipeline(
+            OneDSEventPipelineSettings settings,
+            ILogger logger)
+        {
+            this.settings = settings ?? throw new ArgumentNullException(nameof(settings));
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
```

Comment:
Please replace these two lines with old-fashioned code:

            if (settings == null)
            {
                throw new ArgumentNullException(nameof(settings));
            }
            else
            {
                this.settings = settings;
            }

            if (logger == null)
            {
                throw new ArgumentNullException(nameof(logger));
            }
            else
            {
                this.logger = logger;
            }

to make it compatible with older C# 6.0 (which is used in Unity 2017 (with 4.6 runtime)).